File: module_1/python/day_6/hamming_weight.py
"""
Title: Calculate the Hamming weight.

Problem: Write a function that takes an unsigned integer and return the number
         of '1' bits it has (also known as the Hamming weight).

Execution: python hamming_weight.py
"""
import unittest


def hamming_weight(n: int):
    """Function for calculating the Hamming weight."""
    sum_val = 0
    while n != 0:
        sum_val += 1
        n &= (n - 1)
    return sum_val


class TestHammingWeight(unittest.TestCase):
    """Unit test for Hamming weight."""

    def test_1(self):
        input_val = "00000000000000000000000000001011"
        self.assertEqual(hamming_weight(int(input_val, 2)), 3)
        output = """
        The input binary string 00000000000000000000000000001011 has a total of
        three '1' bits.
        """
        print(f"Explanation: {output}")

    def test_2(self):
        input_val = "00000000000000000000000010000000"
        self.assertEqual(hamming_weight(int(input_val, 2)), 1)
        output = """
        The input binary string 00000000000000000000000010000000has a total of
        one '1' bit.
        """
        print(f"Explanation: {output}")

    def test_3(self):
        input_val = "11111111111111111111111111111101"
        self.assertEqual(hamming_weight(int(input_val, 2)), 31)
        output = """
        The input binary string 11111111111111111111111111111101 has a total of
        thirty-one '1' bits.
        """
        print(f"Explanation: {output}")


if __name__ == '__main__':
    unittest.main()

File: module_1/java/HammingWeight.java
/*
 *   Title: Hamming Weight
 *   Leetcode Link: https://leetcode.com/problems/number-of-1-bits/
 *
 *   Problem: Write a function that takes an unsigned integer and return the
 *   number of '1' bits it has (also known as the Hamming weight).
 *
 *   Input:
 *      int n       => Unsigned integer
 *   Output:
 *      int         => Number of 1 bits in n
 *
 *   Execution: javac HammingWeight.java && java -ea HammingWeight
 */

public class HammingWeight {

    /*
     * Solution #1: Brute force solution using bit shifting
     *
     * Continually shift bits right and check whether the lowest order bit is 1
     *
     * Time Complexity: O(log(n))
     * Space Complexity: O(1)
     */
    public static int hammingWeightBF(int n) {
        int sum = 0;
        while (n > 0) {
            // Check if the lowest-order bit is 1
            if ((n & 1) > 0) sum++;

            // Do an unsigned bitshift
            n >>>= 1;
        }
        return sum;
    }

    /*
     * Solution #2: Optimized solution
     *
     * n & n-1 removes the lowest-order 1 bit in the number. Repeat this until
     * n == 0
     *
     * Time Complexity: O(log(n))  => More optimal when few 1 bits
     * Space Complexity: O(1)
     */
    public static int hammingWeight(int n) {
        int sum = 0;
        while (n != 0) {
            // This removes the lowest-order 1 bit
            n &= n-1;
            sum++;
        }
        return sum;
    }

    public static void main(String[] args) {
        assert hammingWeight(0b1011) == 3;
        assert hammingWeight(0b1000000) == 1;
        assert hammingWeight(0b11111111111111111111111111111101) == 31;

        // ADD TEST CASES HERE

        System.out.println("Passed all test cases");
    }

}
